### ISSUE TITLE
napari works over google chrome remote desktop!

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#886]
File this under potentially interesting rainy day information that doesn't really require any immediate action, but which has implications for remote GL-based rendering.

As has been mentioned [elsewhere](https://github.com/napari/napari/issues/799#issuecomment-565859004), openGL has issues with virtualized desktops, making things like microsoft remote desktop, x11 forwarding on *nix, and (sometimes) tests on our CI fail.  My limited understanding is that the way it is implemented generally limits openGL to older versions (e.g. ≤1.4) that are incompatible with vispy, and [QWindow](https://doc.qt.io/qt-5/qtgui-index.html#opengl-and-opengl-es-integration) rendering using openGL.

Interestingly, napari works just fine using google [chrome remote desktop](https://remotedesktop.google.com/).  I don't really understand why, and googling it hasn't really turned up anything obvious.

Perhaps related (but I'm not really sure), is the [ANGLE library](https://github.com/google/angle) from google which appears to be a layer between OpenGL ES API calls and platform-specific hardware APIs. there is also [discussion about ANGLE in the qt wiki](https://wiki.qt.io/Qt_5_on_Windows_ANGLE_and_OpenGL).

again.  not really sure what to do with that info.  It's also completely possible that other solutions (e.g. teamviewer) would work too and we just haven't tested them yet.  None of this really matters for our more immediate remote goals, but it's just interesting :) and potentially a short-term workaround for some users who really want napari to work in remote desktop environments.

**Comment by jni** [comment]
Well, as much as I hate to create features that are Chrome-only (perhaps there is similar technology from Mozilla?), *ideally* the thing that would be great to do with this information is to figure out how to properly configure a JupyterHub where launching a napari viewer would open a new tab with a virtual napari window in it. Then teaching image analysis in Python using a JupyterHub would be easy!

https://forum.image.sc/t/image-viewer-for-python-notebooks/32976/

(Though note @guiwitz's last response, which I remembered but had never run so I didn't want to propose it as a solution.)

**Comment by sofroniewn** [comment]
@choldgraf - whom I happen to be sitting next to right now :-) curious how you think of the above in terms of napari / jupyterHub integration

**Comment by choldgraf** [comment]
Hey all - a quick thought here, I wonder if the Jupyter Server Proxy would be a way to accomplish this. Basically, if you have a web-app running on a port, then you can proxy that within a Jupyter session and allow users access to it.

A few examples:

* Running tensorboard on Binder: https://github.com/binder-examples/tensorboard
* Running Bokeh on Binder: https://github.com/binder-examples/bokeh
* Running a desktop server in JupyterHub/Binder: https://github.com/yuvipanda/jupyter-desktop-server (this one is more "beta")

**Comment by sofroniewn** [comment]
Thanks for the tips @choldgraf!! Those are all very cool demos.

I wonder if @manics can weigh in as he made this cool demo of napari in binder with a remote desktop https://github.com/manics/jupyter-omeroanalysis-desktop/tree/napari-binder that @guiwitz has used further https://github.com/guiwitz/napari_demo. Does it use that same approach? I think it is based on https://github.com/ryanlovett/nbnovnc which has now been deprecated in favor of https://github.com/yuvipanda/jupyter-desktop-server which was mentioned above.

Just to note here, some of these topics have already been discussed in #495 and there is more info on how we're thinking about remote data / compute more broadly in #881 and what we are currently prioritizing.

**Comment by manics** [comment]
@sofroniewn 

https://github.com/yuvipanda/jupyter-desktop-server is a fork of https://github.com/manics/jupyter-omeroanalysis-desktop/ that provides a generic desktop (so for example my Napari demo could be built on top of it, as could many other desktop apps). They both use jupyter-server-proxy to proxy VNC to the browser.

The more complex explanation: VNC requires TCP traffic which can't be sent direct to a browser. Instead the VNCserver running in Jupyter is wrapped by [websockify](https://github.com/novnc/websockify) which is a library that converts a TCP stream into a websocket that can be handled by JavaScript in the browser. The desktop you see in the browser is [NoVNC](https://github.com/novnc/noVNC), which connects to the websocket.

The VNCserver displays an X application. This doesn't have to be a full desktop, it could be a single window application such as Napari. The problem with the current demo is the Jupyter notebook starts Napari in an existing X session, whereas to get a Napari-only window Napari has to be the root X application. This means either starting Napari directly and connecting to it from Jupyter e.g. using a socket, or having the Jupyter notebook start the whole stack (Napari, websockify, VNCserver). From what I understand #881 may enable the first option? The key requirement is being able to start Napari in one process and connecting to it from a completely separate one.

**Comment by berl** [comment]
resurrecting this issue in case @tlambert03  or @sofroniewn  have seen any progress anywhere on this. Currently I do actually get the napari icon and a napari-named window appearing when I try to open napari on ubuntu remotely via x-forwarding to OS X, but the window is blank and I get lots of 
```WARNING: QOpenGLWidget: Failed to create context
22:43:38 WARNING QOpenGLWidget: Failed to create context
WARNING: QOpenGLWidget: Failed to create context
22:43:38 WARNING QOpenGLWidget: Failed to create context
WARNING: QOpenGLWidget: Failed to create context
22:43:38 WARNING QOpenGLWidget: Failed to create context
WARNING: composeAndFlush: QOpenGLContext creation failed
22:43:39 WARNING composeAndFlush: QOpenGLContext creation failed
WARNING: composeAndFlush: makeCurrent() failed
``` 
which is consistent with the comments above.  I have similar weirdness/failure on remote desktops with xrdp.  

any new ideas here?

**Comment by tlambert03** [comment]
Nope, using virtualgl is still the only working strategy I've seen

**Comment by zbarry** [comment]
Hi all. Just wanted to chime in and say that [NoMachine](https://www.nomachine.com/) has worked for me for running Napari on a remote workstation with a JupyterLab server running on that same box. All I had to do was open up a terminal inside the virtual desktop, get the contents of the `DISPLAY` environment variable, and then plop it into the top of my notebook which I want to spawn Napari from with:

```python
import os
os.environ['DISPLAY'] = ":1000" # (for example)
```

I'm not sure if there's anything special about my NoMachine configuration (could have some org-specific config going on), but maybe this is useful for others looking for alternatives.

**Comment by tlambert03** [comment]
good to know.  yeah I think it may be similar to what google chrome remote desktop is doing.  (which also still works great)

**Comment by berl** [comment]
@zbarry thanks for bringing this back up because I forgot to update:
napari actually works over xrdp now!
machine with napari:
```
ubuntu 20.04 
xrdp 0.9.12
```
remote access with
```
Mac OS 11.4 (Big Sur)
Microsoft Remote Desktop 10.6.7 
```
